In [1]:
%cd ..

D:\SoftUni\Data Science\ProjectV2


### Imports

In [2]:
import pandas as pd

import numpy as np

from src import functions

In this notebook all preprocessed datasets are merged into final working dataset.

### Reading datasets

In [3]:
drugs_distribution_data = pd.read_csv('data/preprocessed_datasets/drugs_distribution_by_country.csv')

In [4]:
drugs_distribution_data

country  heroin_trade  cocaine_trade  \
0                            Turkey           8.0            4.0   
1                        Cabo Verde           4.0            7.0   
2                      South Africa           7.5            5.5   
3                          Colombia           5.0            9.5   
4                              Peru           3.5            9.0   
..                              ...           ...            ...   
188                         Jamaica           3.0            6.0   
189  St. Vincent and the Grenadines           1.0            7.5   
190                        Slovenia           4.0            4.5   
191                       St. Lucia           1.0            7.0   
192                        Dominica           1.0            6.0   

     cannabis_trade  synthetic_drug_trade  
0               5.0                   5.5  
1               5.0                   4.5  
2               3.5                   8.5  
3               8.0                   5.0  
4               5.5                   4.0  
..              ...                   ...  
188             9.0                   3.0  
189             7.5                   1.0  
190             5.0                   5.5  
191             7.0                   1.0  
192             5.0                   2.5  

[193 rows x 5 columns]

In [5]:
mental_disorders_data = pd.read_csv('data/preprocessed_datasets/mental_disorders_by_country.csv')

In [6]:
mental_disorders_data

country  anxiety_disorders_rate  adh_disorders_rate  \
0             Afghanistan                  6036.0              1006.0   
1                  Angola                  4124.0               484.0   
2                 Albania                  4621.0              1014.0   
3                 Andorra                  6651.0              1291.0   
4    United Arab Emirates                  5168.0               448.0   
..                    ...                     ...                 ...   
199                Zambia                  4505.0               487.0   
200              Zimbabwe                  3913.0               485.0   
201                 Egypt                  5214.0              1010.0   
202                 Sudan                  5734.0              1002.0   
203                 China                  3482.0              2184.0   

     biploar_disorders_rate  depressive_disorders_rate  schizophrenia_rate  \
0                     711.0                     5916.0               218.0   
1                     502.0                     6772.0               224.0   
2                     480.0                     2977.0               221.0   
3                     820.0                     4495.0               271.0   
4                     713.0                     4332.0               266.0   
..                      ...                        ...                 ...   
199                   549.0                     4911.0               217.0   
200                   500.0                     3910.0               208.0   
201                   687.0                     4605.0               243.0   
202                   715.0                     5035.0               231.0   
203                   173.0                     2876.0               312.0   

     suds_rate  
0        413.0  
1        277.0  
2        553.0  
3        789.0  
4        732.0  
..         ...  
199      290.0  
200      408.0  
201      368.0  
202      378.0  
203      540.0  

[204 rows x 7 columns]

In [7]:
# suds_data = pd.read_csv('data/preprocessed_datasets/sud_disorders_by_country.csv')

In [8]:
# suds_data

<!-- ### Merging datasets containing data for diffrent mental disorders rates and substance use disorders rates -->

In [9]:
# disorders_data = mental_disorders_data.merge(suds_data, on = 'country', how = 'inner')

In [10]:
# disorders_data

### Preprocessing 'country' column

The goal is to merge 'mental_disorders' and 'drugs distribution' datasets by their common column - 'country'.

In [11]:
countries_in_drugs_data = np.sort(drugs_distribution_data['country'].unique())

In [12]:
len(countries_in_drugs_data)

193

In [13]:
countrues_in_disorders_data = np.sort(mental_disorders_data['country'].unique())

In [14]:
len(countrues_in_disorders_data)

204

Countries in 'mental_disorders_data' are 11 more than in 'drugs_distribution_data'. 
My approach here will be to drop the countries in each dataset which are not present in the another. At the other hand there are some differences between the spelling of the same countries in both datasets, which requires additional prepocessing.

In [15]:
# Country names which are present in drugs distribution dataset but not in the disorders dataset
different_names_in_drugs = np.setdiff1d(countries_in_drugs_data, countrues_in_disorders_data)

In [16]:
different_names_in_drugs

array(['Bolivia', 'Brunei', 'Congo, Dem. Rep.', 'Congo, Rep.',
       'Czech Republic', 'Iran', 'Korea, DPR', 'Korea, Rep.', 'Laos',
       'Liechtenstein', 'Moldova', 'Russia', 'St. Kitts and Nevis',
       'St. Lucia', 'St. Vincent and the Grenadines', 'Syria', 'Tanzania',
       'Turkey', 'United States', 'Venezuela', 'Vietnam'], dtype=object)

In [17]:
# Country names which are present in the disorders dataset but not in the drugs distribution dataset
np.setdiff1d(countrues_in_disorders_data, countries_in_drugs_data)

array(['American Samoa', 'Bermuda', 'Bolivia (Plurinational State of)',
       'Brunei Darussalam', 'Congo', 'Cook Islands', 'Czechia',
       "Democratic People's Republic of Korea",
       'Democratic Republic of the Congo', 'Greenland', 'Guam',
       'Iran (Islamic Republic of)', "Lao People's Democratic Republic",
       'Niue', 'Northern Mariana Islands', 'Palestine', 'Puerto Rico',
       'Republic of Korea', 'Republic of Moldova', 'Russian Federation',
       'Saint Kitts and Nevis', 'Saint Lucia',
       'Saint Vincent and the Grenadines', 'Syrian Arab Republic',
       'Taiwan (Province of China)', 'Tokelau', 'Türkiye',
       'United Republic of Tanzania', 'United States Virgin Islands',
       'United States of America', 'Venezuela (Bolivarian Republic of)',
       'Viet Nam'], dtype=object)

We can observe that data for Liechtenstein as a country is present only in drugs dataframe. Also there are several countries that are present only in mental disorders dataset. The merge process will 'clean' all of them from the resulting dataframe. The only remain for preprocessing countries will be these with different spelled names. A little work of comparing the above arrays is required in order to create an array which contains all different spelled country names in disordes dataset. Then it will be easy to equalize them to the corresponding countries in drugs dataframe and to merge them after.

In [18]:
different_spelled_countries_in_disorders = np.array(['Bolivia (Plurinational State of)', 'Brunei Darussalam', 'Democratic Republic of the Congo', 'Congo', 'Czechia',
                    'Iran (Islamic Republic of)', 'Democratic People\'s Republic of Korea', 'Republic of Korea', 'Lao People\'s Democratic Republic',
                    'Republic of Moldova', 'Russian Federation','Saint Kitts and Nevis', 'Saint Lucia', 'Saint Vincent and the Grenadines',
                    'Syrian Arab Republic', 'United Republic of Tanzania', 'Türkiye','United States of America', 'Venezuela (Bolivarian Republic of)', 'Viet Nam'])

In [19]:
# Delete Liechtenstein in the array
index = np.argwhere(different_names_in_drugs == 'Liechtenstein')
different_names_in_drugs = np.delete(different_names_in_drugs, index)

Now I can create a dictionary with key - value pairs corresponding to all different spelled countries in both dataframes.

In [20]:
different_named_countries_dict = {k: v for k, v in zip(different_spelled_countries_in_disorders, different_names_in_drugs)}

In [21]:
different_named_countries_dict

{'Bolivia (Plurinational State of)': 'Bolivia',
 'Brunei Darussalam': 'Brunei',
 'Democratic Republic of the Congo': 'Congo, Dem. Rep.',
 'Congo': 'Congo, Rep.',
 'Czechia': 'Czech Republic',
 'Iran (Islamic Republic of)': 'Iran',
 "Democratic People's Republic of Korea": 'Korea, DPR',
 'Republic of Korea': 'Korea, Rep.',
 "Lao People's Democratic Republic": 'Laos',
 'Republic of Moldova': 'Moldova',
 'Russian Federation': 'Russia',
 'Saint Kitts and Nevis': 'St. Kitts and Nevis',
 'Saint Lucia': 'St. Lucia',
 'Saint Vincent and the Grenadines': 'St. Vincent and the Grenadines',
 'Syrian Arab Republic': 'Syria',
 'United Republic of Tanzania': 'Tanzania',
 'Türkiye': 'Turkey',
 'United States of America': 'United States',
 'Venezuela (Bolivarian Republic of)': 'Venezuela',
 'Viet Nam': 'Vietnam'}

The dictionary is used to change the misspelled country names in disorders dataset.

In [22]:
functions.rplace_col_values(dict = different_named_countries_dict, df = mental_disorders_data, col = 'country')

D:\SoftUni\Data Science\ProjectV2\src\functions.py:22: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df[col].replace(df[df[col] == key][col].values[0], value, inplace = True)


### Merging disorders and drugs distribution datasets

Now I can merge the data from both dataframes to get a resulted dataframe with 192 observations(number of countries in drugs data - 1 for Liechtenstein) and 11 features.

In [23]:
drugs_disorders_data = drugs_distribution_data.merge(mental_disorders_data, on = "country", how = 'inner')

In [24]:
drugs_disorders_data.head(2)

country  heroin_trade  cocaine_trade  cannabis_trade  \
0      Turkey           8.0            4.0             5.0   
1  Cabo Verde           4.0            7.0             5.0   

   synthetic_drug_trade  anxiety_disorders_rate  adh_disorders_rate  \
0                   5.5                  5420.0              1375.0   
1                   4.5                  3496.0               505.0   

   biploar_disorders_rate  depressive_disorders_rate  schizophrenia_rate  \
0                   737.0                     4730.0               246.0   
1                   473.0                     5801.0               241.0   

   suds_rate  
0      323.0  
1      291.0

In [25]:
drugs_disorders_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 192 entries, 0 to 191
Data columns (total 11 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   country                    192 non-null    object 
 1   heroin_trade               192 non-null    float64
 2   cocaine_trade              192 non-null    float64
 3   cannabis_trade             192 non-null    float64
 4   synthetic_drug_trade       192 non-null    float64
 5   anxiety_disorders_rate     192 non-null    float64
 6   adh_disorders_rate         192 non-null    float64
 7   biploar_disorders_rate     192 non-null    float64
 8   depressive_disorders_rate  192 non-null    float64
 9   schizophrenia_rate         192 non-null    float64
 10  suds_rate                  192 non-null    float64
dtypes: float64(10), object(1)
memory usage: 16.6+ KB


### Saving final dataset

In [26]:
drugs_disorders_data.to_csv('data/final_merged_dataset/drugs_disorders_by_country.csv', index = False)

### Equalize different spelled names in healthcare dataset

In [47]:
healthcare_data = pd.read_csv('data/preprocessed_datasets/healthcare_index_by_country.csv')

In [94]:
healthcare_data

rank      country  healthcare_index
0       1       Taiwan             78.72
1       2  South Korea             77.70
2       3    Australia             74.11
3       4       Canada             71.32
4       5       Sweden             70.73
..    ...          ...               ...
105   106      Bolivia             22.30
106   107     Paraguay             21.90
107   108        Nepal             21.40
108   109     Honduras             19.60
109   110  El Salvador             18.60

[110 rows x 3 columns]

In [48]:
countries_in_healthcare_data = np.sort(healthcare_data['country'].unique())

In [49]:
len(countries_in_healthcare_data)

109

In [50]:
different_names_in_healtcare = np.setdiff1d(countries_in_healthcare_data, np.sort(drugs_disorders_data['country'].unique()))

In [51]:
different_names_in_healtcare

array(['Bosnia And Herzegovina', 'Cote dIvoire',
       'Democratic Republic of the Congo', 'Hong Kong', 'Puerto Rico',
       'South Korea', 'Taiwan'], dtype=object)

In [52]:
np.sort(drugs_disorders_data['country'].unique())

array(['Afghanistan', 'Albania', 'Algeria', 'Andorra', 'Angola',
       'Antigua and Barbuda', 'Argentina', 'Armenia', 'Australia',
       'Austria', 'Azerbaijan', 'Bahamas', 'Bahrain', 'Bangladesh',
       'Barbados', 'Belarus', 'Belgium', 'Belize', 'Benin', 'Bhutan',
       'Bolivia', 'Bosnia and Herzegovina', 'Botswana', 'Brazil',
       'Brunei', 'Bulgaria', 'Burkina Faso', 'Burundi', 'Cabo Verde',
       'Cambodia', 'Cameroon', 'Canada', 'Central African Republic',
       'Chad', 'Chile', 'China', 'Colombia', 'Comoros',
       'Congo, Dem. Rep.', 'Congo, Rep.', 'Costa Rica', 'Croatia', 'Cuba',
       'Cyprus', 'Czech Republic', "Côte d'Ivoire", 'Denmark', 'Djibouti',
       'Dominica', 'Dominican Republic', 'Ecuador', 'Egypt',
       'El Salvador', 'Equatorial Guinea', 'Eritrea', 'Estonia',
       'Eswatini', 'Ethiopia', 'Fiji', 'Finland', 'France', 'Gabon',
       'Gambia', 'Georgia', 'Germany', 'Ghana', 'Greece', 'Grenada',
       'Guatemala', 'Guinea', 'Guinea-Bissau', 'Guyana'

In [39]:
different_spelled_healthcare_countries = {
    'Bosnia And Herzegovina': 'Bosnia and Herzegovina',
    'Cote dIvoire': 'Côte d\'Ivoire',
    'Democratic Republic of the Congo': 'Congo, Dem. Rep.',
    'South Korea': 'Korea, Rep.'
}

In [40]:
functions.rplace_col_values(dict = different_spelled_healthcare_countries, df = healthcare_data, col = 'country')

D:\SoftUni\Data Science\ProjectV2\src\functions.py:22: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df[col].replace(df[df[col] == key][col].values[0], value, inplace = True)


In [41]:
np.sort(healthcare_data['country'].unique())

array(['Albania', 'Algeria', 'Angola', 'Argentina', 'Australia',
       'Austria', 'Azerbaijan', 'Bahrain', 'Bangladesh', 'Belarus',
       'Belgium', 'Bolivia', 'Bosnia and Herzegovina', 'Brazil',
       'Bulgaria', 'Cameroon', 'Canada', 'Chile', 'China', 'Colombia',
       'Congo, Dem. Rep.', 'Costa Rica', 'Croatia', 'Cyprus',
       'Czech Republic', "Côte d'Ivoire", 'Denmark', 'Dominican Republic',
       'Ecuador', 'Egypt', 'El Salvador', 'Estonia', 'Finland', 'France',
       'Georgia', 'Germany', 'Ghana', 'Greece', 'Guatemala', 'Honduras',
       'Hong Kong', 'Hungary', 'Iceland', 'India', 'Indonesia', 'Iran',
       'Iraq', 'Ireland', 'Israel', 'Italy', 'Japan', 'Jordan',
       'Kazakhstan', 'Kenya', 'Korea, Rep.', 'Kuwait', 'Latvia',
       'Lebanon', 'Libya', 'Lithuania', 'Luxembourg', 'Malaysia',
       'Mexico', 'Morocco', 'Myanmar', 'Nepal', 'Netherlands',
       'New Zealand', 'Nigeria', 'North Macedonia', 'Norway', 'Oman',
       'Pakistan', 'Panama', 'Paraguay', 'Peru'

In [56]:
# Countries in durgs_disorders dataset which are not ranked by healthcare index
missing_countries_in_healthcare = np.setdiff1d(drugs_disorders_data['country'].unique(), healthcare_data['country'].unique())

Most of these are island countries, countries with small population and/or least developed or developing countries in Asia and Africa. Their GDP is low, below mode value. [[1]]
Based on low GDP it is very probable these countries to have low level of healthcare systems. I will add the missing countries into the healthcare dataset with a value of 25 percentile of healthcare index.

In [44]:
healthcare_data.describe()

rank  healthcare_index
count  110.000000        110.000000
mean    55.500000         40.888273
std     31.898276         12.945831
min      1.000000         18.600000
25%     28.250000         33.062500
50%     55.500000         38.280000
75%     82.750000         46.140000
max    110.000000         78.720000

In [96]:
additional_healthcare_data = pd.DataFrame({'rank': np.arange(111, 201), 'country': missing_countries_in_healthcare, 'healthcare_index': 33.1})

In [97]:
additional_healthcare_data

rank              country  healthcare_index
0    111          Afghanistan              33.1
1    112              Andorra              33.1
2    113  Antigua and Barbuda              33.1
3    114              Armenia              33.1
4    115              Bahamas              33.1
..   ...                  ...               ...
85   196               Tuvalu              33.1
86   197              Vanuatu              33.1
87   198                Yemen              33.1
88   199               Zambia              33.1
89   200             Zimbabwe              33.1

[90 rows x 3 columns]

In [98]:
healthcare_data_full = pd.concat([healthcare_data, additional_healthcare_data], ignore_index=True)

In [100]:
healthcare_data_full.describe()

rank  healthcare_index
count  200.000000        200.000000
mean   100.500000         37.383550
std     57.879185         10.338574
min      1.000000         18.600000
25%     50.750000         33.100000
50%    100.500000         33.100000
75%    150.250000         39.022500
max    200.000000         78.720000

In [101]:
healthcare_data_full.to_csv('data/preprocessed_datasets/healthcare_index_by_country.csv', index = False)

### References
1. [List of countries by GDP - Wikipedia](https://en.wikipedia.org/wiki/List_of_countries_by_GDP_(nominal))

<a id='references'></a>